src/main/java/com/amihaiemil/docker/ReadLogString.java

```diff
@@ -0,0 +1,208 @@
+ */
+package com.amihaiemil.docker;
+
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.ResponseHandler;
+import org.apache.http.entity.ContentType;
+import org.apache.http.protocol.HTTP;
+import org.apache.http.util.Args;
+import org.apache.http.util.CharArrayBuffer;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.Charset;
+
+/**
+ * Handler that returns the response content as a String.
+ * @author Morozov Evgeniy (lumimul@gmail.com)
+ * @version $Id$
+ * @since 0.0.2
+ */
+final class ReadLogString implements ResponseHandler<String> {
+
+    /**
+     * Handlers to be executed before actually reading the array.
+     */
+    private final ResponseHandler<HttpResponse> other;
+
+    /**
+     * Ctor.
+     * @param other Handlers to be executed before actually reading the array.
+     */
+    ReadLogString(final ResponseHandler<HttpResponse> other) {
+        this.other = other;
+    }
+
+    @Override
+    public String handleResponse(final HttpResponse httpResponse)
+            throws IOException {
+        final HttpResponse resp = this.other.handleResponse(httpResponse);
+        return this.toString(resp.getEntity());
+    }
+
+    /**
+     * Docker logs contains header
+     * [8]byte{STREAM_TYPE, 0, 0, 0, SIZE1, SIZE2, SIZE3, SIZE4}
+     * STREAM_TYPE
+     * 0: stdin (is written on stdout)
+     * 1: stdout
+     * 2: stderr
+     *
+     * SIZE1, SIZE2, SIZE3, SIZE4 are the four bytes of the uint32 size
+     * encoded as big endian.
+     *
+     * This method do:
+     *
+     * 1) Read 8 bytes.
+     * 2) Choose stdout or stderr depending on the first byte.
+     * 3) Extract the frame size from the last four bytes.
+     * 4) Read the extracted size and output it on the correct output.
+     * 5) Goto 1.
+     *
+     * @param entity HttpEntity for read message.
+     * @return Logs from container in String.
+     * @throws IOException if the entity cannot be read
+     */
+    private String toString(final HttpEntity entity) throws IOException {
+        ContentType contentType = ContentType.get(entity);
+        InputStream instream = entity.getContent();
+        int capacity = this.getCapacity(entity);
+        CharArrayBuffer buffer = new CharArrayBuffer(capacity);
+        if (instream != null) {
+            try {
+                Charset charset = this.getCharset(contentType);
+                Reader reader = new InputStreamReader(instream, charset);
+                this.read(buffer, reader);
+            } finally {
+                instream.close();
+            }
+        }
+        return buffer.toString();
+    }
+
+    /**
+     * Read data from reader and save it in buffer.
+     *
+     * @param buffer Buffer for save message.
+     * @param reader Reader for read message.
+     * @throws IOException if the entity cannot be read
+     */
```

Comment:
@lumimul Can you please add a more detailed JavaDoc? Please specify exactly what this method is doing. It contains a very complicated algorithm that nobody will understand later.

Reply:
add more detail)